src/rvsdg/mod.rs

```diff
@@ -127,6 +144,10 @@ pub(crate) enum RvsdgBody {
 /// to nodes by their index in the vector.
 pub struct RvsdgFunction {
     /// The number of input arguments to the function.
+    ///
+    /// Functions all take `n_args + 1` arguments, where the last argument is a
+    /// "state edge" used to preserve ordering constraints to (potentially)
+    /// impure function calls.
     pub(crate) n_args: usize,
```

Comment:
Should you include the `+ 1` in n_args so that later we can have functions that don't necessarily take this argument?

Reply:
I prefer the existing structure but I'm open to changing it. I think having the state edge be initialized to `Arg(n_args)` is nicer than `Arg(n_args-1)`, because you're implicitly assuming that n_args (which is a `usize`) is at least 1.

I agree that it'd be nice to later have functions that don't necessarily take this argument (e.g. when we read them back out of egglog and we've proved that it's pure).

But in that case, I don't think it's a property of the function, it's the fact that _calls_ to the function won't have any state edges in them. In either case, we don't have to change this struct. WDYT?

Reply:
That sounds like a good argument but I'm a little confused still.
Does that mean that we can call a function without the state argument, but the function still takes that argument implicitly? seems like an arity mismatch

Reply:
@ezrosent I'm still unclear on this

Reply:
Sorry; I didn't revisit this page after merging (in the future, I'd be happy to try and resolve these threads before merging the PR).

> Does that mean that we can call a function without the state argument, but the function still takes that argument implicitly? seems like an arity mismatch

Right now, all functions will take a state argument, so whether or not it counts in `n_args` or not is just a matter of convention.

If we remove state arguments as part of optimizations in egglog, then there won't be any mention of `Arg(n_args)` in the body and we will have to make sure not to extract any calls to this function that pass a state edge by mistake.

Those are all real concerns, but I don't see them as being harder or easier with `n_args` being what it is here. Maybe that'll become clearer once we extract more things.

I don't feel strongly about this though; I'm fine changing it.

Reply:
Oh, I see
Yeah I see what you are saying, I think it doesn't matter except that it's confusing that you have the same `n_args` whether you are passing the state edge or not.